wazo_auth/services/saml.py

```
@@ -117,57 +123,148 @@ def prepare_redirect_response(
+        ).decode()
+        req_id, info = client.prepare_for_authenticate(relay_state=relay_state)
+
+        self._outstanding_requests[req_id] = SamlAuthContext(
             saml_session_id,
             redirect_url,
             domain,
+            relay_state,
         )
         location = [i for i in info['headers'] if i[0] == 'Location'][0][1]
         return location, saml_session_id
 
+    def _decode_saml_response(
+        self, saml_client: Saml2Client, saml_response: str, conv_info: dict[str, Any]
+    ) -> None | AuthnResponse:
+        return saml_client.parse_authn_request_response(
+            saml_response,
+            BINDING_HTTP_POST,
+            self._outstanding_requests,
+            None,
+            conv_info=conv_info,
+        )
+
+    def _find_session_by_relay_state(
+        self, relay_state: str
+    ) -> tuple[SamlAuthContext, SamlAuthContext] | tuple[None, None]:
+        session: list[SamlAuthContext] = [
+            e
+            for e in self._outstanding_requests
+            if self._outstanding_requests[e].relay_state == relay_state
+        ]
+        if len(session) == 1:
+            return self._outstanding_requests[session[0]], session[0]
+        else:
+            logger.warning(
+                "Unable to get SAML session corresponding to the received RelayState"
+            )
+            return None, None
+
+    def _process_auth_response_error(
+        self, redirect_url: str, req_id: Any, msg: str
+    ) -> None:
+
+        logger.warning(msg)
+        logger.debug(f'Removing session: {req_id}')
+        del self._outstanding_requests[req_id]
+        raise exceptions.SAMLProcessingErrorWithReturnURL(
+            'Unknown principal', return_url=redirect_url
+        )
+
+    def _process_auth_response_context_not_found(self) -> None:
+        logger.warning('ACS response request failed: Context not found')
+        raise exceptions.SAMLProcessingError('Context not found', code=404)
```

Not sure I see the value of encapsulating a log and an exception. They would benefit from being embedded in the context where they should occur. 
Hiding exceptions behind method calls leads to code being harder to read(you don't know by seeing a method call that this is actually unconditionally raising an exception).